From 933ffcf2166e795e031a66f653a344f5127d6cba Mon Sep 17 00:00:00 2001
From: "qix.a.zhang" <qix.a.zhang@intel.com>
Date: Mon, 21 Mar 2016 16:33:37 +0800
Subject: [PATCH 38/41] usb: gadget: f_fs: avoid data lost when dequeue

When the usb io operation is interrupted by signal, wait for
another 100ms to avoid data lost.

Change-Id: Ie38d5e724f28c7dd85b64807f1d2f5f790fa97a4
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-31683
Signed-off-by: Zhi Jin <zhi.jin@intel.com>
Signed-off-by: Zhang Di <di.zhang@intel.com>
Signed-off-by: Marcin Walewski <marcin.walewski@intel.com>
Reviewed-on: https://android.intel.com:443/484032
Reviewed-on: https://android.intel.com:443/494176
Reviewed-on: https://android.intel.com:443/504785
---
 drivers/usb/gadget/f_fs.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/usb/gadget/f_fs.c b/drivers/usb/gadget/f_fs.c
index 6e63461..8e3fccc 100644
--- a/drivers/usb/gadget/f_fs.c
+++ b/drivers/usb/gadget/f_fs.c
@@ -678,6 +678,10 @@ static ssize_t ffs_epfile_io(struct file *file,
 			goto out;
 
 		if (unlikely(wait_for_completion_interruptible(&done))) {
+			/* Wait for another 100ms to avoid data lost */
+			if (wait_for_completion_timeout(&done, HZ / 10))
+				goto done;
+
 			/*
 			 * To avoid race condition with ffs_epfile_io_complete,
 			 * dequeue the request first, then check status.
@@ -688,6 +692,7 @@ static ssize_t ffs_epfile_io(struct file *file,
 			usb_ep_dequeue(ep->ep, req);
 			interrupted = true;
 		}
+done:
 		/*
 		 * XXX We may end up silently droping data here.
 		 * Since data_len (i.e. req->length) may be bigger
-- 
1.9.1

